File: TD/03-tdp-dicho-knapsack/src/knap/Knapsack.java
public class Knapsack {

    /** Les données du problème,
     * On copie ici plutot que de les passer en paramètre de la récursion, 
     * puisqu'ils ne changent pas au cours de la récursion. */
    private int [] valeur;
    private int capacite;
    
    /** La meilleure solution connue pour l'instant. 
     * La fonction récursive a (entre autres) des paramètres du même type 
     * qu'elle modifie lorsqu'elle construit sa solution. Lorsqu'une solution
     * meilleure que la meilleure connue est trouvée, elle est stockée ici.
     */
    private KnapsackSolution meilleure;
    
    
    /* Constructeur */
    public Knapsack(int c, int [] vals){
	capacite = c;
	valeur = vals;

	meilleure = new KnapsackSolution(vals);	
    }
    
    /* quelques Getter pour des attributs privés */
    public int getMeilleureValeur(){return meilleure.getValeur();}
    public boolean [] getMeilleureSolution(){return meilleure.getPris();}


    /* change l'objet en chaine de caractères pour le visualiser. Pratique pour debugger */
    public String toString(){
	return toString(valeur.length);
    }  

    /* idem, mais n'affiche que les premiers objets (ie, une solution partielle) */
    public String toString(int objMax) {
	String s = "";
	s = s + "\nValeurs:  ";
	for (int i=0; i<objMax; i++)
	    s = s + valeur[i] + " ";
	s = s + meilleure.toString();
	s = s + "\nCapacité: " + capacite;
	return s;
    }
 
    private void chercheRec(int profondeur, KnapsackSolution courante) throws Exception {
	System.out.print("(n="+profondeur+") Explore "+courante.toString(profondeur));
	if (courante.getValeur() > capacite) {
	    System.out.println("  *** Oups, ca deborde (backtrack!) ***");
	    return;
	}

	if (courante.getValeur() > meilleure.getValeur()) {
	    System.out.print("   Nouvelle meilleure solution ");
	    meilleure = courante.duplique();
	} else {
	    System.out.print("    ");
	}
       
	if (profondeur == valeur.length) {
	    System.out.println("(Cas terminal)");
	    return;
	} else {       
	    System.out.println("(Cas général)");
	}
       
       
       /*
	if (courante.getValeur() == capacite) {
	    System.out.println("   *** Solution parfaite ***");
	    throw new Exception();
	}*/

	/* Prend l'objet et récurse */
	courante.prendObjet(profondeur);
	chercheRec(profondeur+1, courante);

	/* Pose l'objet et récurse */
	courante.poseObjet(profondeur);
	chercheRec(profondeur+1, courante);
    }

    public void cherche() {
	try {
	    chercheRec(0, new KnapsackSolution(valeur));
	} catch (Exception e) {}
    }
}

File: TD/03-tdp-dicho-knapsack/scala/knapsack.scala
/** Problème du sac à dos */

object knapsack { // KILL

/* ************************* */
/* Diverses fonctions d'aide */
/* ************************* */

/** Calcule le gain d'un sac à dos donné */
def valeurTotale(objets:Array[Boolean], poids:Array[Int]):Int = {
  var total = 0
  for (i <- 0 to objets.length -1)
    if (objets(i)) 
      total += poids(i)
  return total
}


/** Affiche si les objets du sac à dos sont pris ou non dans une solution partielle (en cours de construction)
 *
 * Le second argument est la taille déja construite.
 */
def affiche(objets:Array[Boolean], objetMax:Int) {
	      
  // on utilise un min pour se protéger du cas (fréquent) où cette
  // fonction est appellée avec un second argument supérieur à la
  // taille du tableau.
  for (i <- 0 to math.min(objetMax, objets.length-1))
    if (objets(i))
      print(" O ")
    else
      print(" N ")
  if (objetMax < objets.length-1)
    for (i <- objetMax to objets.length-2)
      print("...")
  print(";")
}


/** Dupplique un sac dans un autre, pour se souvenir de la meilleure solution connue */
def dupplique(src:Array[Boolean], dst:Array[Boolean]) {
  for (i <- 0 to src.length-1) 
    dst(i) = src(i)
}

/** indique dans le paramètre que l'objet spécifié est maintenant pris
    (cette fonction est juste là pour se simplifier la vie) */
def prendObjet(objets:Array[Boolean], obj:Int) {
    if (objets(obj)) 
        println("L'objet "+obj+" est déjà pris; ignore la requete");
    objets(obj) = true
}
/** indique dans le paramètre que l'objet spécifié est maintenant posé */
def poseObjet(objets:Array[Boolean], obj:Int) {
    if (!objets(obj)) 
        println("L'objet "+obj+" est déjà posé; ignore la requete");
     objets(obj) = false
}

/* ********************** */
/* La fonction principale */
/* ********************** */

/** La fonction publique, pour chercher la meilleure solution */
def cherche(poids:Array[Int] , capacite:Int) {

  // on va beaucoup utiliser cette valeur, alors on fait un alias poursimplifier les écritures suivantes
  val len = poids.length
  
  // Affiche l'instance du problème
  print("Poids des objets: ")
  for (i <- 0 to poids.length-1) 
    print(" "+poids(i)+" ")
  println("; Capacite: "+capacite)
  
  // variable locale pour sauvegarder la meilleure solution connue à tout moment
  var meilleure:Array[Boolean] = Array.fill(len)(false)

  // Initialise l'appel récursif
  chercheRec(0,  // On commence le remplissage au rang 0
             Array.fill(len)(false) // Cette écriture crée un tableau
                                    // dans lequel la valeur 'false'
                                    //est répétée 'len' fois.
  ) // fin des paramètres de chercheRec. La récursion est lancée.
  
  // Affiche la meilleure solution trouvée
  println
  print("Meilleure solution trouvée:")
  for (i <- 0 to meilleure.length-1)
    print(" "+poids(i)+":"+(if (meilleure(i)) "O" else "N")+";")
  println(" Valeur:"+valeurTotale(meilleure,poids)+" (la capacité était "+capacite+")")
  
  
  // Défini l'appel récursif
  def chercheRec(profondeur:Int, courante:Array[Boolean]) {

    val valeur = valeurTotale(courante,poids)
    
    print(" (prof="+profondeur+") Explore ")
    affiche(courante,profondeur)
    print(" Valeur: "+valeur)
    
    if (valeur > capacite) {
      println(" *** Oups, ca deborde (backtrack!) ***")
      return
    }
    if (valeurTotale(courante, poids) > valeurTotale(meilleure, poids)) {
      dupplique(courante,meilleure)
      print(" Nouvelle meilleure solution ");
    } else {
      print("   ")
    }
    if (profondeur == poids.length) {
      println("(Cas terminal)");
      return;
    } else {       
      System.out.println("(Cas général)");
    }
						           
					
    /* Prend l'objet et récurse */
    prendObjet(courante,profondeur);
    chercheRec(profondeur+1, courante);
	
    /* Pose l'objet et récurse */
    poseObjet(courante, profondeur);
    chercheRec(profondeur+1, courante);
				
  }
}


def main(args:Array[String]) {
  if (args.length == 0) {
    println("Usage: knapsack <capacite> <obj1> <obj2> ... <objN>")
    println("Arguments par défaut: 10 5 3 2")
    cherche(Array(5,4,3,2), 10)
    
  } else {  
    val capa = args(0).toInt
    val objets:Array[Int] = new Array(args.length-1)
    for (i <- 0 to objets.length-1)
      objets(i) = args(i+1).toInt
    cherche(objets, capa)  
  }
}

}
